corehq/apps/couch_sql_migration/rewind.py

```diff
@@ -0,0 +1,130 @@
+import logging
+import re
+import sys
+from datetime import datetime, timedelta
+
+import attr
+
+from couchforms.models import XFormInstance
+
+from corehq.util.couch_helpers import NoSkipArgsProvider
+from corehq.util.pagination import ResumableFunctionIterator
+
+from .asyncforms import get_case_ids
+
+log = logging.getLogger(__name__)
+
+
+def rewind_iteration_state(statedb, domain, move_to):
+    doc_type = "XFormInstance"
+    status_interval = timedelta(minutes=1)
+    next_status = datetime.utcnow() - status_interval
+    rw = Rewinder(statedb, domain, doc_type, move_to)
+    if rw.offset:
+        offset = rw.offset
+        for n, item in enumerate(rw, start=1):
+            if n > offset:
+                rw.save_state(item)
+                break
+            if datetime.utcnow() > next_status:
+                log.info("scanned %s docs", n)
+                next_status = datetime.utcnow() + status_interval
+
+
+@attr.s
+class Rewinder:
+    statedb = attr.ib()
+    domain = attr.ib()
+    doc_type = attr.ib()
+    move_to = attr.ib()
+
+    def __attrs_post_init__(self):
+        for method, regex in [
+            ("case_rewind", r"^case-(\d+)$"),
+            ("resume_rewind", r"^resume-"),
+        ]:
+            match = re.search(regex, self.move_to)
+            if match:
+                getattr(self, method)(match)
+                break
+        else:
+            raise NotImplementedError(self.move_to)
+        migration_id = self.statedb.unique_id
+        resume_key = "%s.%s.%s" % (self.domain, self.doc_type, migration_id)
+        self.itr = ResumableFunctionIterator(resume_key, None, None, None)
+
+    def resume_rewind(self, match):
+        self.offset = None
```

Comment:
Is `self.offset` an initial position, and `self.move_to` the rewound position we want to go to? "Offset" and "move to" could be synonyms, but I'm not sure what names would be clearer.

I know that `__init__` calls `__attrs_post_init__` and `__attrs_post_init__` calls this, but it still feels to me not to be an obvious place to create and initialize attributes. Would it be clearer to add this at the start of the class? e.g.
```python
    offset = attr.ib(init=False)
```

I can't see where `Rewinder` uses it. What is it for?

Reply:
This should probably be explained better in a docstring or something, although I view most of this as throw-away code (hopefully it won't be used too often). It's used to indicate the rewind scanner offset:

https://github.com/dimagi/commcare-hq/blob/9df5c98e938f20a67d900aef6ba0c42cf7035add/corehq/apps/couch_sql_migration/rewind.py#L23-L26

It is used by `case_rewind` to hold the integer offset parsed form `self.move_to` string. It's set to `None` here in `resume_rewind` so it evaluates to false, causing the forms/cases scanning feature to be disabled.

`self.move_to` is a user-supplied string. Different things happen depending on its format. For example: 
- `case-1000`: scan back through 1000 previously iterated cases
- `resume-2019-09-20T14:23:14.706428Z`: set the main forms iteration state to a previously found point/state, either the original state of the iteration prior to initiating a `case-...` rewind (this is written into the statedb at the beginning of each case rewind), or the point found and written into the statedb after a `case-...` rewind has concluded.